### PR TITLE
Add support for variable length extension metadata blocks

### PR DESCRIPTION
**Pull request by quietvoid**
Implements fixes for #104, #105

L8, L9, and L10 metadata blocks can have variable lengths, to avoid serializing unnecessary metadata.
That's useful in the case of allowing preset mastering display primaries, as well as being able to support custom primaries.

### CODE REVIEW

**Comment by quietvoid** [comment]
@saindriches I added your changes, just need to actually test it and fix stuff.

File: dolby_vision/src/rpu/extension_metadata/blocks/level9.rs [hunk 1]
```diff
@@ -34,10 +83,62 @@ impl ExtMetadataBlockInfo for ExtMetadataBlockLevel9 {
...
     fn bytes_size(&self) -> u64 {
-        1
+        self.length
     }
 
     fn required_bits(&self) -> u64 {
-        8
+        match self.length {
+            1 => 8,
+            9 => 136,
```
Comment (saindriches) [inline dolby_vision/src/rpu/extension_metadata/blocks/level9.rs:92]: Should be `17 => 136,`

File: dolby_vision/src/rpu/extension_metadata/blocks/level9.rs [hunk 2]
```diff
@@ -34,10 +83,62 @@ impl ExtMetadataBlockInfo for ExtMetadataBlockLevel9 {
...
+#[cfg(feature = "serde_feature")]
+impl Serialize for ExtMetadataBlockLevel9 {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let name = "ExtMetadataBlockLevel9";
+        let fields_count = match self.length {
+            1 => 1,
+            9 => 9,
```
Comment (saindriches) [inline dolby_vision/src/rpu/extension_metadata/blocks/level9.rs:122]: Also here

File: src/dovi/tests.rs [hunk 3]
```diff
@@ -432,17 +432,21 @@ fn cmv40_full_rpu() -> Result<()> {
...
                 ..Default::default()
             }),
             ExtMetadataBlock::Level9(ExtMetadataBlockLevel9 {
+                length: 1,
                 source_primary_index: 0,
+                ..Default::default()
             }),
             ExtMetadataBlock::Level10(ExtMetadataBlockLevel10 {
-                target_display_index: 255,
+                target_display_index: 21,
```
Comment (saindriches) [inline src/dovi/tests.rs:445]: `target_display_index: 21` is predefined, this causes error

File: src/dovi/tests.rs [hunk 4]
```diff
@@ -852,3 +856,91 @@ fn generate_full_hdr10plus() -> Result<()> {
...
+                anchor_pq: 0,
+                anchor_power: 0,
+            }),
+            ExtMetadataBlock::Level8(ExtMetadataBlockLevel8 {
+                length: 25,
+                target_display_index: 123,
+                ..Default::default()
+            }),
+            ExtMetadataBlock::Level9(ExtMetadataBlockLevel9 {
+                length: 1,
```
Comment (saindriches) [inline src/dovi/tests.rs:899]: Should be `length: 17,` here

**Comment by quietvoid** [comment]
Feels a bit odd that the primaries are over 15 bits, since the documentation uses values based on 50 000 max.

**Comment by quietvoid** [comment]
Looks like there's no way to validate L10.
However L9 passes the tests:
```
Test case results (' - ' = not applicable or skipped by configuration):
Test case Id    1: FAIL (Video codec compliance)
Test case Id    2:   -  (Frame continuity)
Test case Id    3: PASS (VUI validation)
Test case Id    4:   -  (Dual layer validation)
Test case Id    5: PASS (RPU NAL unit presence and position)
Test case Id    6: PASS (RPU header compliance)
Test case Id    7: PASS (Basic DM metadata ('L0') - intrinsic)
Test case Id    8: PASS (Basic DM metadata ('L0') - source comparison)
Test case Id    9: PASS (L1 metadata - intrinsic)
Test case Id   10: PASS (L1 metadata - source comparison)
Test case Id   11: PASS (L2 metadata - source comparison)
Test case Id   12: FAIL (L4 metadata presence)
Test case Id   13: PASS (L6 metadata presence and stability)
Test case Id   14: PASS (Comparison with reference RPU)
Test case Id   15: PASS (L5 metadata presence)
Test case Id   16: PASS (Extended display management metadata general validation)
Test case Id   17: PASS (L3 metadata - source comparison)
Test case Id   18: PASS (L8 metadata - intrinsic)
Test case Id   19: PASS (L8 metadata - source comparison)
Test case Id   20: PASS (L9 metadata - intrinsic)
Test case Id   21: PASS (L9 metadata - source comparison)
Test case Id   22: PASS (L10 metadata - intrinsic)
Test case Id   23:   -  (ST 2094-10 syntax check)
Test case Id   24:   -  (ST 2094-10 ext_block_level 1)
Test case Id   25:   -  (ST 2094-10 ext_block_level 2)
Test case Id   26:   -  (ST 2094-10 ext_block_level 5)
```

**Comment by saindriches** [comment]
It will raise an error if L8 and L10 trims do not correspond (that means any `target_display_id` which is predefined or not used by L8 trims also should not appear in L10, and any target in L8 which is not predefined must appear in L10), so it should be OK.

**Comment by saindriches** [comment]
It's able to verify L10 with an XML metadata, it checks target displays used by L8 and compare correspond L10 with XML data, and will raise an error if values don't match.

Use `cmv4_0_2_custom_displays.xml` with changed values in `TargetDisplay`:

```
Level 8 Display Management Metadata mismatch of RPU and external source reference file [E243].
Found Level 10 for target_display_index 255 in RPU but no legal corresponding custom target display in reference XML/MXF.
```

It will also raise an error if L8 and L10 trims do not correspond (that means any `target_display_id` which is predefined or not used by L8 trims also should not appear in L10, and any target in L8 which is not predefined must appear in L10), so it should be OK.

File: dolby_vision/src/xml/parser.rs [hunk 5]
```diff
@@ -558,30 +704,181 @@ impl CmXmlParser {
...
+                .all(|(a, b)| (*a - *b).abs() < f64::EPSILON)
+        });
+
+        // Exact match to preset primaries
+        let primary_index = if let Some((primary_index, _)) = matching_primaries {
+            if is_source {
+                primary_index
+            } else {
+                // FIXME: Why are the target primaries offset by the preset source primaries?
+                primary_index + level9::PREDEFINED_COLORSPACE_PRIMARIES.len()
```
Comment (quietvoid) [inline dolby_vision/src/xml/parser.rs:837]: I still don't understand why this is.
Probably going to have to experiment with different primary values.

Comment (quietvoid) [reply dolby_vision/src/xml/parser.rs:839]: Anyways, this is starting to get pretty big so I'd rather merge before fixing more.

Comment (saindriches) [reply dolby_vision/src/xml/parser.rs:839]: Sorry for late, L10 should only use the `PREDEFINED_COLORSPACE_PRIMARIES`, while the `PREDEFINED_REALDEVICE_PRIMARIES` part is only for L9 from my test. That means L9 may have a `target_primary_index: 18` but L10 doesn’t.

If a target display uses primaries from the real devices list, then verifier do not recognize a `target_primary_index` derived from that list as predefined, we should use 255 in that moment.

Comment (quietvoid) [reply dolby_vision/src/xml/parser.rs:839]: I'll have to play around with both primary/source index values.
If there's an issue with the current implementation, can create a new issue.

**Pull request merged**